### ISSUE TITLE
Proposal for a new more reliable per Episode identifier - Episode ID, GUID hashing and such

### ISSUE THREAD

**Issue by brianoflondon** [Podcastindex-org/podcast-namespace#186]
When working on and thinking about the Value for Value system I think it's essential we have a cross system unique identifier for each episode.

I thought the episode ID stored in PodcastIndex would take that place but around the release of Episode 21 of Podcasting 2.0 my hack script failed because I relied on it. When Adam corrected his feed to reverse the incorrect posting order of Episode 21/20 the index gave a new ID to the episode. Adam didn't change the GUID:

```
        {
            "id": 1526276331,
            "title": "Episode 20: Mapping the Road",
            "link": "http:\/\/adam.curry.com\/html\/PC202020210115Podcas-KkVnRt1g8X7hZq2NLKqznxdWCTfw8w.html",
            "guid": "PC2020",
```

However I know that the original episode ID was 1472557573 (because I wrote that as [part of the permalink into this post on Hive](https://peakd.com/hive-136933/@podcastindex/episode-20-mapping-the-road---1472557573):

```https://peakd.com/hive-136933/@podcastindex/episode-20-mapping-the-road---1472557573```

[That episode ID no longer is in the Podcastindex](https://api.podcastindex.org/api/1.0/episodes/byid?id=1472557573&pretty)

```
{"status":"true","episode":[],"count":0,"id":"1472557573","description":"Episode not found."}
```

My new code gave me ```076480ba``` for the next episode:

```https://peakd.com/hive-136933/@podcastindex/episode-22-this-train-is-rolling-076480ba```

So I've changed my algo for generating my unique, per-episode permalinks to combining the Title of the episode and the first 8 characters MD5 Hash of the GUID. I'm wondering if that hashed GUID, or perhaps a combined hash of Feed name and GUID might be something we can all use?

```
        guid = epi.get('guid')
        if guid == '':
            guid ='in the morning'
        m = hashlib.md5()
        m.update(guid.encode())
        hashstr = m.hexdigest()[:8]
        p_link = epi['title'] + '_' + hashstr
        p_link = sanitize_permlink(p_link)
```

This does mean my script will re-post after a title change and I'm still thinking about the correct behaviour there but that's my issue.

**Comment by brianoflondon** [comment]
Oh just by the by, Episode 20 of Podcasting 2.0 is in the index twice:

https://api.podcastindex.org/api/1.0/episodes/byid?id=1526276330&pretty

```{
    "status": "true",
    "id": "1526276330",
    "episode": {
        "id": 1526276330,
        "title": "Episode 21: Hooray for Hollywood",
        "link": "http:\/\/adam.curry.com\/html\/PC202120210122Podcas-ZBQLkDB7XBMdbSzHgz9vpRJ04J4dkT.html",
        "description": "<p><b>Shownotes<\/b><\/p><p>\n<b>Podcasting 2.0 for January 22nd 2021 Episode 21: Hooray for Hollywood<\/b><\/p><p>\nThe man who can memorize more transactions than the mempool - Dave Jones!<\/p><p>\nCrank those value sliders up everyone!<\/p><p>\nDownload the mp3<\/p><p>\nCheck out the podcasting 2.0 apps and services newpodcastapps.com<\/p><p>\nSupport us with your Time Talent and Treasure<\/p><p>\nCheck out all the 2.0 apps and hosting companies: newpodcastapps.com<\/p><p>\n<b>Shownotes<\/b><\/p><p>\nAdam & Dave discuss the week's developments on podcastindex.org<\/p><p>\nBlubrry<\/p><p>\nTodd Cochrane...",
        "guid": "PC2021",
        "datePublished": 1611345628,
        "datePublishedPretty": "January 22, 2021 2:00pm",
        "dateCrawled": 1611454102,
        "enclosureUrl": "https:\/\/mp3s.nashownotes.com\/PC20-21-2021-01-22-Final.mp3",
        "enclosureType": "audio\/mpeg",
        "enclosureLength": 66505501,
        "duration": 0,
        "explicit": 0,
        "episode": null,
        "episodeType": null,
        "season": 0,
        "image": "https:\/\/noagendaassets.com\/enc\/1601061118.678_pciavatar.jpg",
        "feedItunesId": null,
        "feedImage": "https:\/\/noagendaassets.com\/enc\/1601061118.678_pciavatar.jpg",
        "feedId": 920666,
        "feedTitle": "Podcasting 2.0",
        "feedLanguage": "en",
        "chaptersUrl": "https:\/\/studio.hypercatcher.com\/chapters\/podcast\/http:mp3s.nashownotes.compc20rss.xml\/episode\/PC2021",
        "transcriptUrl": "https:\/\/mp3s.nashownotes.com\/PC20-21-Captions.srt"
    },
    "description": "Found matching item."
}```

https://api.podcastindex.org/api/1.0/episodes/byid?id=1526276331&pretty

```{
    "status": "true",
    "id": "1526276331",
    "episode": {
        "id": 1526276331,
        "title": "Episode 20: Mapping the Road",
        "link": "http:\/\/adam.curry.com\/html\/PC202020210115Podcas-KkVnRt1g8X7hZq2NLKqznxdWCTfw8w.html",
        "description": "<p><b>Shownotes<\/b><\/p><p>\n<b>Podcasting 2.0 for January 15th 2021 Episode 20: Mapping the Road<\/b><\/p><p>\nThe man who can bake an endpoint that changes humanity before breakfast - Dave Jones!<\/p><p>\nCrank those value sliders up everyone!<\/p><p>\nDownload the mp3<\/p><p>\nCheck out the podcasting 2.0 apps and services newpodcastapps.com<\/p><p>\nSupport us with your Time Talent and Treasure<\/p><p>\nCheck out all the 2.0 apps and hosting companies: newpodcastapps.com<\/p><p>\n<b>Shownotes<\/b><\/p><p>\nAdam & Dave discuss the week's developments on podcastindex.org<\/p><p>\nPeople search<\/p><...",
        "guid": "PC2020",
        "datePublished": 1610741600,
        "datePublishedPretty": "January 15, 2021 2:13pm",
        "dateCrawled": 1611454102,
        "enclosureUrl": "https:\/\/mp3s.nashownotes.com\/PC20-20-2021-01-15-Final.mp3",
        "enclosureType": "audio\/mpeg",
        "enclosureLength": 79226089,
        "duration": 0,
        "explicit": 0,
        "episode": null,
        "episodeType": null,
        "season": 0,
        "image": "https:\/\/noagendaassets.com\/enc\/1601061118.678_pciavatar.jpg",
        "feedItunesId": null,
        "feedImage": "https:\/\/noagendaassets.com\/enc\/1601061118.678_pciavatar.jpg",
        "feedId": 920666,
        "feedTitle": "Podcasting 2.0",
        "feedLanguage": "en",
        "chaptersUrl": "https:\/\/studio.hypercatcher.com\/chapters\/podcast\/http:mp3s.nashownotes.compc20rss.xml\/episode\/PC2020",
        "transcriptUrl": "https:\/\/mp3s.nashownotes.com\/PC20-20-Captions.srt"
    },
    "description": "Found matching item."
}
```

**Comment by MartinMouritzen** [comment]
I am in full support of using the GUID more across the api, instead of the podcastindex episode id.

However I do not see the point of using anything from the title (which I imagine could change - where the GUID should be permanent)

**Comment by brianoflondon** [comment]
> However I do not see the point of using anything from the title (which I imagine could change - where the GUID should be permanent)

Yes, using the title is just something I'm doing for my use. 

But thinking about it, we might need to combine show name or id and guid in a hash to avoid collisions if someone is using something like an incrementing counter for guid.

**Comment by brianoflondon** [comment]
Having thought about my own code I decided to combine the PodcastIndex ID and the GUID together (with an _) in order to get the blog that I hash and I've bumped up to taking the first 10 chars of the MD5 hash. 

I'll be using this schema tonight for the next No Agenda and going forward.

```
        pod_info_id = podInfo.get('feed').get('id') 
        guid = epi.get('guid')
        if guid == '':
            guid ='in the morning'
        id_and_guid = f'{pod_info_id}_{guid}'
        m = hashlib.md5()
        m.update(id_and_guid.encode())
        hashstr = m.hexdigest()[:10]
        p_link = epi['title'] + '_' + hashstr
        p_link = sanitize_permlink(p_link)
```

**Comment by dellagustin** [comment]
Welcome to the Jungle.

The brutal thing about RSS is that absolutely nothing is 100% reliable as unique identifier.
Per spec, the guid should be the one tag to use as unique id within a feed, but many podcasters do not understand what it is, how to use it, the consequences of not using, etc...

When I first coded podStation I relied on guid being part of the feed, only to later discover it was not mandatory, and had to a major refactoring including some heuristc to identify an episode.

Relying on feed url is also problematic (athough I do that, due to the lack of anything else better), as it is not uncommon for podcasters to not own the url, and eventually change it.

In summary, I think feedurl + guid is the best we can get, albeit far from perfect, with a fall back to feedurl + enclosureurl in case guid is unavailable Sometimes I think that pubdate and title are less likely to change than enclosureurl, but only actual data and statistics could prove that.

**Comment by brianoflondon** [comment]
Anyone got any thoughts on first 10 chars of an MD5 hash? Is that sane? I've not got a lot of experience on how prone that would be to collision problems.

**Comment by agates** [comment]
@brianoflondon that would be a mess for collisions, especially as md5 is one of the worst for that.

I've got some ideas for this problem but I'm going to sit on them for a while until I think them through.  My primary question for @daveajones is why did the episode ID change?  And is that code available somewhere?

**Comment by brianoflondon** [comment]
I'd be delighted if PodcastIndex was the keeper of the unique IDs but I think we do need to know how they're allocated.

**Comment by saerdnaer** [comment]
I would prefer following mechanim: If the guid is already a uuid, than directly use it. If it is anything else use uuid_v5() as hashing function to get one. 

If you prefer a shorter string you can encode it in base62 or simular.

https://en.wikipedia.org/wiki/Universally_unique_identifier#Versions_3_and_5_(namespace_name-based)

**Comment by brianoflondon** [comment]
> I would prefer following mechanim: If the guid is already a uuid, than directly use it. If it is anything else use uuid_v5() as hashing function to get one.
> 
> If you prefer a shorter string you can encode it in base62 or simular.
> 
> https://en.wikipedia.org/wiki/Universally_unique_identifier#Versions_3_and_5_(namespace_name-based)

The basic requirement is that we have something we can all independently generate on the fly based on information in the feed. If that same info is ALSO in the podcastindex that makes it even better.

The issue comes because we have no idea how @daveajones is making the Episode ID and why it would change when Adam got the Pub date wrong and changed it without changing the Episode GUID or anything else about it.

**Comment by swschilke** [comment]
A real GUID / UUID is the thing to go, the RSS description is at the moment "something which is unique" (in this RSS) which doesn't make it unique across all RSS files.Some people just use the filename

**Comment by agates** [comment]
Thinking about this more, I would just use uuid_v5 with the URL namespace for the enclosure or link URL.

GUID is such a bad name for this.  Thanks, Microsoft.

**Comment by yoshimo** [comment]
I would not only go for an episode level identifier but for a show unique identifier.
This would help for master feeds that have more than one show . We would be able to split them apart in the clients UI in case the user wants to be more specific

**Comment by daveajones** [comment]
> I would not only go for an episode level identifier but for a show unique identifier.
> This would help for master feeds that have more than one show . We would be able to split them apart in the clients UI in case the user wants to be more specific

Each

**Comment by saerdnaer** [comment]
@daveajones I think we can close this one in favour of https://github.com/Podcastindex-org/podcast-namespace/issues/251

**Comment by saerdnaer** [comment]
@brianoflondon Does the mechanism documented at https://github.com/Podcastindex-org/podcast-namespace/blob/main/docs/other-recommendations.md#episode-guid solve your issues or do we still need a better way to achieve this? I want to especially focus on the last sentence:
> Consumers of podcast feeds should fall back to the enclosure URL or the namespaced UUIDv5 (SHA1 Hash with ns:URL) of the enclosure URL.

Which would roughly be: 

```python
from uuid import UUID, uuid5, NAMESPACE_URL
from urllib.parse import urlparse

# example item
episode = {    
    "id": 1526276331,
    "title": "Episode 20: Mapping the Road",
    "link": "http://adam.curry.com\/html\/PC202020210115Podcas-KkVnRt1g8X7hZq2NLKqznxdWCTfw8w.html",
    "guid": "PC2020",
    "enclosure": { "url": "https://mp3s.nashownotes.com/PC20-20-2021-01-15-Final.mp3" }
}
episode_uuid = None
if episode.guid:
  try:
      episode_uuid = UUID(episode.guid)
  except ValueError:
    try:
      # if guid value is an URL, and use NAMESPACE_URL for uuid v5 generation
      # we could also use the isPermaLink from `<guid isPermaLink="false">PC2020</guid>` but can not really trust it…
      urlparse(episode.guid)
      episode_uuid = uuid5(NAMESPACE_URL, "https://mp3s.nashownotes.com/PC20-56-2021-10-01-Final.mp3")
  except ValueError:
      # the value is some random string, generated UUID from that:
      episode_uuid = uuid5(PODCAST_UUID, episode.guid)
else:  
  # empty guid value: fallback to the enclosure URL
  if episode.enclosure.url:
    episode_uuid = uuid5(NAMESPACE_URL, episode.enclosure.url)
```

Or if you want to to the last part with some online generator:
https://www.uuidtools.com/generate/v5
<img width="969" alt="image" src="https://user-images.githubusercontent.com/40266/135714216-8c2aed96-4476-4a00-b332-d3cc3b1e35c4.png">

Or should we drop the protocol from the enclosure URL and use 
the Podcast Namspace UUID or the UUID value from the `<podcast:guid>` tag instead? c.f. https://github.com/Podcastindex-org/podcast-namespace/issues/251#issuecomment-843534772
```python
from uuid import UUID, uuid5

PODCAST_UUID = UUID('{value from <podcast:guid>}')  instead?
uuid5(PODCAST_UUID, "mp3s.nashownotes.com/PC20-56-2021-10-01-Final.mp3")
```

@daveajones We should probably also add the `podcastGuid` or `uuid` attributes to the API's episode responses (e.g. https://podcastindex-org.github.io/docs-api/#get-/episodes/byfeedid) and add an `By GUID` route like we have for podcast feeds.

---

## Episode GUID

For all new episodes, we recommended to use

* either a permanent URI, e.g. `https://example.com/ep0003`
* or a [Universally unique identifier (UUID)](https://en.wikipedia.org/wiki/Universally_unique_identifier) e.g. `7c029615-a810-5214-9342-eee73f58435d`

The GUID of an episode should never change, even if a new version of the episode is being published, otherwise this episode will be duplicated downstream. **Consumers of podcast feeds should fall back to the enclosure URL or the namespaced UUIDv5 (SHA1 Hash with ns:URL) of the enclosure URL.**


#### Examples

Only one entry per `<item>` is valid:
```xml
<guid>https://example.com/ep0003</guid>
<guid isPermaLink="false">7c029615-a810-5214-9342-eee73f58435d</guid>
```
(`isPermaLink` is optional, its default value is `true`)

**Comment by saerdnaer** [comment]
Can some summarize the current state of this topic?
